src/main/scala/calc/Compiler.scala

```diff
@@ -61,21 +65,89 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.flatMap((x:Tree) => findCapture(x)).toSet
+      case Call(f, args) => findCapture(f) ++ args.flatMap((x:Tree) => findCapture(x)).toSet
+    }
+  }
+
+  var seed = 1
+  def generateName():String = {
+    seed += 1
+    "generated_"+seed.toString
+  }
+
+
   /** Compile an expression tree into an IR `Tree`, which is an expression
    *  that evaluates to the result of the tree.
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
+  def compileExpr(tree: Tree)(env:Map[String, irt.Tree] = Map(),typeEnv:Map[String, irtpe.Type] = Map()): irt.Tree = {
     implicit val pos = tree.pos
 
     tree match {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) =>
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs)(env, typeEnv), compileExpr(rhs)(env, typeEnv))
+
+      case Let(ident, value, body) =>
+        val jsV = compileExpr(value)(env, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, mutable = false, jsV)
+        irt.Block(List(jsId, compileExpr(body)(env + (ident.name -> jsV), typeEnv + (ident.name -> jsV.tpe))))
```

Comment:
It is a code smell that you still have `env` at all. `typeEnv` is necessary, of course, but `env` should never be needed, since the value tree should only ever appear as the rhs of the `VarDef`, but in the body you should only use the newly declared variable.